File: viewModel/viewModel/app/src/main/java/com/example/viewmodel/MainActivity.java
package com.example.viewmodel;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProviders;

import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

public class MainActivity extends AppCompatActivity {

    Button button,button1;
    TextView textView;
    Integer scoreTillNow;
    scoreViewModel ScoreViewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        textView = findViewById(R.id.textView);
        button = findViewById(R.id.button);
        button1 = findViewById(R.id.button1);

        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                increaseByOne();
            }
        });

        button1.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                reset();
            }
        });
        ScoreViewModel = ViewModelProviders.of(this).get(scoreViewModel.class);
        textView.setText(String.valueOf(ScoreViewModel.getScore()));

      //  scoreTillNow=0;
        //textView.setText(String.valueOf(scoreTillNow));
    }

    public void increaseByOne(){
      ScoreViewModel.addScore();
        textView.setText(String.valueOf(ScoreViewModel.getScore()));
      //  scoreTillNow++;
        // textView.setText(String.valueOf(scoreTillNow));
    }

    public void reset(){
        ScoreViewModel.resetScore();
        textView.setText(String.valueOf(ScoreViewModel.getScore()));
       // scoreTillNow=0;
        //textView.setText(String.valueOf(scoreTillNow));
    }
}

File: orientationChange/app/src/main/java/com/example/orientationchange/MainActivity.kt
package com.example.orientationchange

import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity

class MainActivity : AppCompatActivity() {
    lateinit var decrement: Button
    lateinit var increment: Button
    lateinit var textView: TextView
    private var num = 0
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        decrement = findViewById(R.id.decrement)
        increment = findViewById(R.id.increment)
        textView = findViewById(R.id.textview)
        decrement.setOnClickListener(View.OnClickListener {
            num--
            if (num < 0) {
                num = 0
            }
            textView.setText(num.toString())
        })
        increment.setOnClickListener(View.OnClickListener {
            num++
            textView.setText(num.toString())
        })

       /* if (savedInstanceState != null) {
            num = savedInstanceState.getInt("count");
//            textView.setText(String.valueOf(num));
            textView.setText(num.toString())
        }*/
        //either this or line 47 to 51
    }

    //saving the instance
    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putInt("count", num)
    }

    //geeting the value of the saved instance
    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        num = savedInstanceState.getInt("count")
        textView!!.text = num.toString()
    }
}
